yarn-project/prover-client/src/prover/bb_prover.ts

```diff
@@ -69,98 +103,112 @@ export class BBNativeRollupProver implements CircuitProver {
     logger.info(`Using native BB at ${config.bbBinaryPath} and working directory ${config.bbWorkingDirectory}`);
     logger.info(`Using native ACVM at ${config.acvmBinaryPath} and working directory ${config.acvmWorkingDirectory}`);
 
-    const mappings = await BBNativeRollupProver.generateVerificationKeys(config);
-
-    return new BBNativeRollupProver(config, mappings);
-  }
-
-  public static async generateVerificationKeys(bbConfig: BBProverConfig) {
```

Comment:
Are keys generated on demand?

Later edit: yep, see `getVerificationKeyDataForCircuit` that fills up the promise cache

Reply:
Yeah, verification keys are now written out whenever we generate a proof. This is a bit inefficient as we re-create the same vk everytime we run the circuit. But creating them at startup is also inefficient as most of the work in creating a vk is first creating the proving key, which we are doing anyway when we generate the proof.

The ultimate solution will be determined by the characteristics of Honk and what that looks like in terms of it's keys and proof generation times. It might be that we run a long running process, or maybe we can read the proving keys in from file or something. All of this key caching stuff is subject to change.